### ISSUE TITLE
[BUG] When sending a heartbeat, throw KeeperException$NoNodeException

### ISSUE THREAD

**Issue by SunJiFengPlus** [apache/incubator-dolphinscheduler#2921]
**Describe the bug**
problem in ```OSUtils::getHost```,
``` java
    try {
      return InetAddress.getLocalHost().getHostAddress();
    } catch (UnknownHostException e) {
      logger.error(e.getMessage(),e);
    }
    return null;
```

sometimes ```InetAddress.getLocalHost().getHostAddress();``` return 127.0.0.1, sometimes return other address, when using other address sending a heartbeat to zookeeper, throw ```KeeperException$NoNodeException```. <br>

---

问题在 ```OSUtils::getHost``` 这个方法中, ```InetAddress.getLocalHost().getHostAddress(); ```这个方法有时候返回的是 127.0.0.1, 但是有时候会返回其他IP. 当以其他IP发送心跳时候就会抛 ```KeeperException$NoNodeException```<br>

**To Reproduce**
i dont kwon how to reproduce, but it really happened on my mac,  i just start application, after a while, application throw ```KeeperException$NoNodeException```.  On debug mode, i find problem in ```OSUtils::getHost```<br>

---

不太清楚如何在其他电脑上复现, 但确实发生在我的Mac上, 启动应用后过一会就会抛 ```KeeperException$NoNodeException```. 在Debug的时候我定位到问题出在 ```OSUtils::getHost``` <br>

**Which version of Dolphin Scheduler:**
```<version>1.2.1-SNAPSHOT</version>```

**Context:**
my hosts file:
``` shell
##
# Host Database
#
# localhost is used to configure the loopback interface
# when the system is booting.  Do not change this entry.
##
127.0.0.1	localhost
255.255.255.255	broadcasthost
::1             localhost
# BEGIN section for OpenVPN Client SSL sites
127.94.0.1	client.openvpn.net
# END section for OpenVPN Client SSL sites
```

**Comment by CalvinKirs** [comment]
> **Describe the bug**
> problem in `OSUtils::getHost`,
> 
> ```java
>     try {
>       return InetAddress.getLocalHost().getHostAddress();
>     } catch (UnknownHostException e) {
>       logger.error(e.getMessage(),e);
>     }
>     return null;
> ```
> 
> sometimes `InetAddress.getLocalHost().getHostAddress();` return 127.0.0.1, sometimes return other address, when using other address sending a heartbeat to zookeeper, throw `KeeperException$NoNodeException`.
> 
> 问题在 `OSUtils::getHost` 这个方法中, `InetAddress.getLocalHost().getHostAddress(); `这个方法有时候返回的是 127.0.0.1, 但是有时候会返回其他IP. 当以其他IP发送心跳时候就会抛 `KeeperException$NoNodeException`
> 
> **To Reproduce**
> i dont kwon how to reproduce, but it really happened on my mac, i just start application, after a while, application throw `KeeperException$NoNodeException`. On debug mode, i find problem in `OSUtils::getHost`
> 
> 不太清楚如何在其他电脑上复现, 但确实发生在我的Mac上, 启动应用后过一会就会抛 `KeeperException$NoNodeException`. 在Debug的时候我定位到问题出在 `OSUtils::getHost`
> 
> **Which version of Dolphin Scheduler:**
> `<version>1.2.1-SNAPSHOT</version>`
> 
> **Context:**
> my hosts file:
> 
> ```shell
> ##
> # Host Database
> #
> # localhost is used to configure the loopback interface
> # when the system is booting.  Do not change this entry.
> ##
> 127.0.0.1	localhost
> 255.255.255.255	broadcasthost
> ::1             localhost
> # BEGIN section for OpenVPN Client SSL sites
> 127.94.0.1	client.openvpn.net
> # END section for OpenVPN Client SSL sites
> ```

Hello, about the other ip you said, is this ip configured in your host file?

**Comment by yangyichao-mango** [comment]
> **Describe the bug**
> problem in `OSUtils::getHost`,
> 
> ```java
>     try {
>       return InetAddress.getLocalHost().getHostAddress();
>     } catch (UnknownHostException e) {
>       logger.error(e.getMessage(),e);
>     }
>     return null;
> ```
> 
> sometimes `InetAddress.getLocalHost().getHostAddress();` return 127.0.0.1, sometimes return other address, when using other address sending a heartbeat to zookeeper, throw `KeeperException$NoNodeException`.
> 
> 问题在 `OSUtils::getHost` 这个方法中, `InetAddress.getLocalHost().getHostAddress(); `这个方法有时候返回的是 127.0.0.1, 但是有时候会返回其他IP. 当以其他IP发送心跳时候就会抛 `KeeperException$NoNodeException`
> 
> **To Reproduce**
> i dont kwon how to reproduce, but it really happened on my mac, i just start application, after a while, application throw `KeeperException$NoNodeException`. On debug mode, i find problem in `OSUtils::getHost`
> 
> 不太清楚如何在其他电脑上复现, 但确实发生在我的Mac上, 启动应用后过一会就会抛 `KeeperException$NoNodeException`. 在Debug的时候我定位到问题出在 `OSUtils::getHost`
> 
> **Which version of Dolphin Scheduler:**
> `<version>1.2.1-SNAPSHOT</version>`
> 
> **Context:**
> my hosts file:
> 
> ```shell
> ##
> # Host Database
> #
> # localhost is used to configure the loopback interface
> # when the system is booting.  Do not change this entry.
> ##
> 127.0.0.1	localhost
> 255.255.255.255	broadcasthost
> ::1             localhost
> # BEGIN section for OpenVPN Client SSL sites
> 127.94.0.1	client.openvpn.net
> # END section for OpenVPN Client SSL sites
> ```

Can you provide the other IP you said?

`InetAddress.getLocalHost().getHostAddress();` actually is not a system independent reliable method, and it will encounter problems in multi network environment.

The system will register the worker and master nodes in zk, and if your network environment change, maybe it will cause the `KeeperException$NoNodeException`.

**Comment by SunJiFengPlus** [comment]
``` 2020-06-08 07:36:13.512 ERROR 17592 --- [tBeatExecutor_1] o.a.d.service.zk.ZookeeperOperator       : update key : /dolphinscheduler/nodes/worker/default/127.94.0.1:1234 , value : 0.27,0.66,1.75,2020-06-07 18:56:35,2020-06-08 07:36:13 ```
Another situation return ``` 127.94.0.1 ```,  This is the configuration in my hosts file

**Comment by yangyichao-mango** [comment]
The reason for the IP change is the `InetAddress.getLocalHost().getHostAddress()` will change when your `InetAddress.addressCache.cache.xxx.value.addresses`'s first value change, and the result of `InetAddress.getLocalHost().getHostAddress()` equals your `InetAddress.addressCache.cache.xxx.value.addresses`'s first value.

The result when I connect the wifi:
![image](https://user-images.githubusercontent.com/29545877/83985785-080fbd00-a96d-11ea-9224-7f6ac2b5ce59.png)

The result when I turn off the network:
![image](https://user-images.githubusercontent.com/29545877/83985977-bca9de80-a96d-11ea-98ee-9b774826fcf1.png)

So when it returns `127.94.0.1` not `127.0.0.1`, it maybe your network env changed. 

Can you show the results of `ifconfig` on the command line of your computer and the result of the picture above when your network env changed?

Although there is little possibility of IP changes in the production environment, but I think we should compare the results of zk record and the result of `InetAddress.getLocalHost().getHostAddress()` to determine the node?

@dailidong Can you help to confirm this question? Thx.

**Comment by SunJiFengPlus** [comment]
This is an ifconfig result without desensitization. I hope it helps.

```
lo0: flags=8049<UP,LOOPBACK,RUNNING,MULTICAST> mtu 16384
	options=1203<RXCSUM,TXCSUM,TXSTATUS,SW_TIMESTAMP>
	inet 127.0.0.1 netmask 0xff000000
	inet6 ::1 prefixlen 128
	inet6 fe80::1%lo0 prefixlen 64 scopeid 0x1
	inet 127.94.0.1 netmask 0xff000000
	nd6 options=201<PERFORMNUD,DAD>
gif0: flags=8010<POINTOPOINT,MULTICAST> mtu 1280
stf0: flags=0<> mtu 1280
en0: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=400<CHANNEL_IO>
	ether a4:5e:60:d7:89:bd
	inet6 fe80::106c:8a72:4c45:2478%en0 prefixlen 64 secured scopeid 0x4
	inet 172.17.1.202 netmask 0xfffffc00 broadcast 172.17.3.255
	nd6 options=201<PERFORMNUD,DAD>
	media: autoselect
	status: active
p2p0: flags=8843<UP,BROADCAST,RUNNING,SIMPLEX,MULTICAST> mtu 2304
	options=400<CHANNEL_IO>
	ether 06:5e:60:d7:89:bd
	media: autoselect
	status: inactive
awdl0: flags=8943<UP,BROADCAST,RUNNING,PROMISC,SIMPLEX,MULTICAST> mtu 1484
	options=400<CHANNEL_IO>
	ether 4a:db:8e:c7:2d:11
	inet6 fe80::48db:8eff:fec7:2d11%awdl0 prefixlen 64 scopeid 0x6
	nd6 options=201<PERFORMNUD,DAD>
	media: autoselect
	status: active
llw0: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=400<CHANNEL_IO>
	ether 4a:db:8e:c7:2d:11
	inet6 fe80::48db:8eff:fec7:2d11%llw0 prefixlen 64 scopeid 0x7
	nd6 options=201<PERFORMNUD,DAD>
	media: autoselect
	status: active
en1: flags=8963<UP,BROADCAST,SMART,RUNNING,PROMISC,SIMPLEX,MULTICAST> mtu 1500
	options=460<TSO4,TSO6,CHANNEL_IO>
	ether 82:13:05:e4:90:c0
	media: autoselect <full-duplex>
	status: inactive
en2: flags=8963<UP,BROADCAST,SMART,RUNNING,PROMISC,SIMPLEX,MULTICAST> mtu 1500
	options=460<TSO4,TSO6,CHANNEL_IO>
	ether 82:13:05:e4:90:c1
	media: autoselect <full-duplex>
	status: inactive
bridge0: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=63<RXCSUM,TXCSUM,TSO4,TSO6>
	ether 82:13:05:e4:90:c0
	Configuration:
		id 0:0:0:0:0:0 priority 0 hellotime 0 fwddelay 0
		maxage 0 holdcnt 0 proto stp maxaddr 100 timeout 1200
		root id 0:0:0:0:0:0 priority 0 ifcost 0 port 0
		ipfilter disabled flags 0x0
	member: en1 flags=3<LEARNING,DISCOVER>
	        ifmaxaddr 0 port 8 priority 0 path cost 0
	member: en2 flags=3<LEARNING,DISCOVER>
	        ifmaxaddr 0 port 9 priority 0 path cost 0
	nd6 options=201<PERFORMNUD,DAD>
	media: <unknown type>
	status: inactive
utun0: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 1380
	inet6 fe80::ee58:2e5d:c91c:f75%utun0 prefixlen 64 scopeid 0xb
	nd6 options=201<PERFORMNUD,DAD>
utun1: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 2000
	inet6 fe80::a49:c23c:8187:be02%utun1 prefixlen 64 scopeid 0xc
	nd6 options=201<PERFORMNUD,DAD>
utun2: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 1380
	inet6 fe80::5628:2952:d245:4c77%utun2 prefixlen 64 scopeid 0xd
	nd6 options=201<PERFORMNUD,DAD>
utun3: flags=8051<UP,POINTOPOINT,RUNNING,MULTICAST> mtu 1380
	inet6 fe80::808:197:eabb:939%utun3 prefixlen 64 scopeid 0xe
	nd6 options=201<PERFORMNUD,DAD>
```

**Comment by CalvinKirs** [comment]
The original method may return a different IP during operation.fix to #2924.
